src/stream/src/executor/sync_kv_log_store.rs

```diff
@@ -0,0 +1,1045 @@
+//!    For any subsequently flushed chunks, we will read them via
+//!    `flushed_chunk_future`. See the next paragraph below.
+//!
+//!    We will next read `flushed_chunk_future` (if there's one pre-existing one), see below for how
+//!    it's constructed, what it is.
+//!
+//!    Finally we will pop the earliest item in the buffer.
+//!    - If it's a chunk yield it.
+//!    - If it's a watermark yield it.
+//!    - If it's a flushed chunk reference (`LogStoreBufferItem::Flushed`),
+//!      we will read the corresponding chunks in the log store.
+//!      This is done by constructing a `flushed_chunk_future` which will read the log store
+//!      using the `seq_id`.
+//!   - Barrier,
+//!     because they are directly propagated from the upstream when polling it.
+//!
+//! TODO(kwannoel):
+//! - [] Add dedicated metrics for sync log store, namespace according to the upstream.
+//! - [] Add tests
+//! - [] Handle watermark r/w
+//! - [] Handle paused stream
+
+use std::collections::VecDeque;
+use std::pin::Pin;
+
+use await_tree::InstrumentAwait;
+use futures::future::BoxFuture;
+use futures::{FutureExt, TryStreamExt};
+use futures_async_stream::try_stream;
+use itertools::Itertools;
+use risingwave_common::array::StreamChunk;
+use risingwave_common::bitmap::{Bitmap, BitmapBuilder};
+use risingwave_common::catalog::{TableId, TableOption};
+use risingwave_common::hash::VnodeBitmapExt;
+use risingwave_common_estimate_size::EstimateSize;
+use risingwave_connector::sink::log_store::{ChunkId, LogStoreResult};
+use risingwave_hummock_sdk::table_watermark::{VnodeWatermark, WatermarkDirection};
+use risingwave_storage::store::{
+    InitOptions, LocalStateStore, NewLocalOptions, OpConsistencyLevel, SealCurrentEpochOptions,
+};
+use risingwave_storage::StateStore;
+use tokio::select;
+
+use crate::common::log_store_impl::kv_log_store::buffer::LogStoreBufferItem;
+use crate::common::log_store_impl::kv_log_store::reader::timeout_auto_rebuild::TimeoutAutoRebuildIter;
+use crate::common::log_store_impl::kv_log_store::reader::{
+    read_flushed_chunk, read_persisted_log_store,
+};
+use crate::common::log_store_impl::kv_log_store::serde::{
+    KvLogStoreItem, LogStoreItemMergeStream, LogStoreRowSerde,
+};
+use crate::common::log_store_impl::kv_log_store::{
+    FlushInfo, KvLogStoreMetrics, KvLogStoreReadMetrics, ReaderTruncationOffsetType, SeqIdType,
+    FIRST_SEQ_ID,
+};
+use crate::executor::prelude::*;
+use crate::executor::{
+    Barrier, BoxedMessageStream, Message, StreamExecutorError, StreamExecutorResult,
+};
+
+type StateStoreStream<S> = Pin<Box<LogStoreItemMergeStream<TimeoutAutoRebuildIter<S>>>>;
+type ReadFlushedChunkFuture = BoxFuture<'static, LogStoreResult<(ChunkId, StreamChunk, u64)>>;
+
+struct SyncedKvLogStoreExecutor<S: StateStore> {
+    actor_context: ActorContextRef,
+    table_id: TableId,
+    read_metrics: KvLogStoreReadMetrics,
+    metrics: KvLogStoreMetrics,
+    serde: LogStoreRowSerde,
+    seq_id: SeqIdType,
```

Comment:
Fields like `seq_id`, `truncation_offset`, `flushed_chunk_future` and `buffer` can be removed from the executor struct, and instead declare as variables in `execute_inner`.